arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,264 @@
+        }
+      } else if (isFileNameSst(*it)) {
+        std::unordered_map<std::string, std::string>::const_iterator hashIt;
+        {
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          hashIt = _fileNamesToHashes.find(*it);
+        }
+        if (hashIt == _fileNamesToHashes.end()) {
+          std::string tempPath =
+              basics::FileUtils::buildFilename(_rootPath, *it);
+          LOG_TOPIC("d6c86", DEBUG, arangodb::Logger::ENGINES)
+              << "checkMissingShaFiles:"
+                 " Computing checksum for "
+              << tempPath;
+          auto checksumCalc = ChecksumCalculator();
+          if (TRI_ProcessFile(tempPath.c_str(),
+                              [&checksumCalc](char const* buffer, size_t n) {
+                                checksumCalc.updateEVPWithContent(buffer, n);
+                                return true;
+                              })) {
+            checksumCalc.computeFinalChecksum();
+            writeShaFile(tempPath, checksumCalc.getChecksum());
+          }
+        }
+      }
+    }
+  }
+}
+
+std::string ChecksumHelper::removeFromTable(std::string const& fileName) {
+  std::string checksum;
+  {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    if (auto it = _fileNamesToHashes.find(TRI_Basename(fileName));
+        it != _fileNamesToHashes.end()) {
+      checksum = it->second;
+      _fileNamesToHashes.erase(it);
+    }
+  }
+  return checksum;
+}
+
+std::string ChecksumHelper::buildShaFileNameFromSst(
+    std::string const& fileName, std::string const& checksum) {
+  if (!fileName.empty() && !checksum.empty()) {
+    TRI_ASSERT(fileName.size() > 4);
+    std::string shaFileName = fileName.substr(0, fileName.size() - 4);
+    TRI_ASSERT(!isFileNameSst(shaFileName));
+    shaFileName += ".sha." + checksum + ".hash";
+    return shaFileName;
+  }
+  return {};
+}
+
+rocksdb::Status ChecksumWritableFile::Append(const rocksdb::Slice& data) {
+  _checksumCalc.updateIncrementalChecksum(data.data(), data.size());
+  return rocksdb::WritableFileWrapper::Append(data);
+}
+
+rocksdb::Status ChecksumWritableFile::Close() {
+  TRI_ASSERT(_helper != nullptr);
+  _checksumCalc.computeFinalChecksum();
+  if (!_helper->writeShaFile(_fileName, _checksumCalc.getChecksum())) {
+    return rocksdb::Status::Corruption("File writing was unsuccessful");
+  }
+  return rocksdb::WritableFileWrapper::Close();
+}
+
+rocksdb::Status ChecksumEnv::NewWritableFile(
+    const std::string& fileName, std::unique_ptr<rocksdb::WritableFile>* result,
+    const rocksdb::EnvOptions& options) {
+  std::unique_ptr<rocksdb::WritableFile> writableFile;
+  rocksdb::Status s =
+      rocksdb::EnvWrapper::NewWritableFile(fileName, &writableFile, options);
+  if (!s.ok()) {
+    return s;
+  }
+  try {
+    if (_helper->isFileNameSst(fileName)) {
+      result->reset(
+          new ChecksumWritableFile(std::move(writableFile), fileName, _helper));
```

Comment:
We should avoid using `new` and dealing with raw pointer. Can you use `std::make_unique` here?